keps/sig-network/3015-prefer-local-traffic-policy/README.md

```diff
@@ -0,0 +1,535 @@
+deprecate the old name?
+
+<<[/UNRESOLVED]>>
+```
+
+[public documentation on externalTrafficPolicy]: https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip
+
+### Internal Traffic Policy
+
+This is easy; source IP is always preserved for internal traffic, so
+`PreferLocal` can be implemented in the obvious way with no caveats.
+
+```
+<<[UNRESOLVED internal-traffic-policy-local ]>>
+
+Given `PreferLocal`, is there any reason to keep the behavior of
+`internalTrafficPolicy: Local` with no fallback? Is there actually a
+use case for "either send locally or drop the traffic"?
+
+The KEP gives the use case:
+
+    As a platform owner, I want to create a Service that always
+    directs traffic to a logging daemon on the same node. Traffic
+    should never bounce to a daemon on another node.
+
+But there's an implied "and if the logging daemon is unavailable, then
+logs should be dropped". Are there really cases where you want that?
+And the answer can't be "the administrator will just make sure that
+the logging daemon is always available", because (a) you can't do
+that, and (b) if you could do that then `PreferLocal` would give the
+same result as `Local` anyway, since it would never need to fall back.
+
+The strongest argument for keeping distinct `Local` and `PreferLocal`
+in the internal traffic case is that it makes it parallel with the
+external traffic case (other than the source IP stuff). But I feel
+like we really need to document "you almost never want
+`internalTrafficPolicy: Local`".
+
+If we rename `externalTrafficPolicy: Local` then there's a stronger
+argument for deprecating `internalTrafficPolicy: Local` along with it.
+
+<<[/UNRESOLVED]>>
+```
+
+### Interaction with Topology
+
+The [Topology Aware Hints KEP] says "Both ExternalTrafficPolicy and
+InternalTrafficPolicy will be given precedence over topology aware
+routing." (By which it really means that _`Local`_ traffic policy has
+precedence over topology, and topology only applies to `Cluster`
+traffic policy.)
+
+When delivering packets locally, `PreferLocal` policy will have to
+ignore topology, just like `Local` policy does.
+
+When delivering packets remotely, it makes the most sense to have
+`PreferLocal` policy make use of topology. This means a `PreferLocal`
+service always either sends to the same endpoints it would have sent
+to if it was `Local`, or to the same endpoints it would have sent to
+if it was `Cluster`; there's no need to add a third case.
+
+(FWIW, the expected situation when `PreferLocal` has to fall back to
+sending to a remote endpoint is when pods are in flux, and some nodes
+that are "supposed to" have an endpoint don't yet/anymore. But when
+that happens, there's a good chance that the topology code will _also_
+hit its fallback case (due to there being at least one zone which is
+missing an endpoint), and so topology won't even get used anyway. At
```

Comment:
I'm not sure this is true. For example, there may be a Service with 3 endpoints spread over 6 nodes. That would mean "Local" would fallback to "Cluster" ~1/2 of the time, but topology would consistently be helpful. Having a "PreferLocal" mode may result in a larger set of users than the traditional "Local" use cases, including some cases where users are not intending to deploy workloads on every possible node.

Reply:
> there may be a Service with 3 endpoints spread over 6 nodes. That would mean "Local" would fallback to "Cluster" ~1/2 of the time

(I assume you meant `PreferLocal` would fall back to `Cluster`.)

> including some cases where users are not intending to deploy workloads on every possible node.

Hm... I was not considering that as a use case, although it does make sense. It's the same behavior, you're just thinking of it as "`Cluster` with optimization" rather than "`Local` with fallback". I should add a User Story for that.

At any rate, the comment here about `PreferLocal` and Topology failing at the same time was purely "informational" and doesn't affect the actual behavior; the proposed rule ("use a local endpoint if you can, otherwise do exactly what you would have done if the policy was `Cluster`") will give the right answer whether or not Topology is in effect.